driver-core/src/main/com/mongodb/internal/connection/Cluster.java

```diff
@@ -89,4 +85,17 @@ void selectServerAsync(ServerSelector serverSelector, OperationContext operation
      * Server Discovery And Monitoring</a> specification.
      */
     void onChange(ServerDescriptionChangedEvent event);
+
+    /**
+     * A non-atomic snapshot of the servers in a {@link Cluster}.
+     */
+    @ThreadSafe
+    interface ServersSnapshot {
```

Comment:
Implementing `ServersSnapshot` the way it's done now, allows to see clearly in the diff how the transition from `getServer` to `getServersSnapshot` did not change the behavior of getting a server.

We could have two different implementations: one containing a `Map`, and one containing a `Server`, but I can't see myself why this approach would be objectively better than the current one.

> the hashmap is being copied outside of the lambda - so I don't see how we can guarantee this is ThreadSafe

Could you please explain in more details what problem with the current implementation you see?

Reply:
- We talked over Zoom - there is no problem with the current implementation. There was still an idea to have a single implementation of `ServersSnapshot` that contains a `Map<ServerAddress, Server>`. However, such an implementation turned out to be impossible: `LoadBalancedCluster.server` does not have the `ServerAddress` information, which means, the method `LoadBalancedCluster.getServersSnapshot` can't create a `Map<ServerAddress, Server>`. When I agreed [here](https://github.com/mongodb/mongo-java-driver/pull/1367#discussion_r1594765696) that an assertion would have made sense, I didn't realize that the `ServerAddress` information is not available via `ClusterableServer`.

- There was also a concern that someone may in the future accidentally start using `addressToServerTupleMap` from the lambda expression that is currently in the method `AbstractMultiServerCluster.getServersSnapshot`, which will be incorrect. But the same statement is valid about any place where we write a lambda expression that can technically capture and use state it is not supposed to use; it is also valid about instance method that may technically use the instance state it is not supposed to use.

Reply:
That is too bad about the assertion. The main impediment I see now is that ServerTuple is private, and I see no clean way around that. How about the following change to docs?

> Provides access to a (non-atomic) snapshot of the servers in a {@link Cluster}, mapped by their address.

Reply:
I think, you advocated that we should attempt to document program elements that declare types as the things they are, as opposed to what they represent/provide/etc. For example, _"A boolean value ..."_ for `MqlBoolean` as opposed to "Represents a boolean value ...". This may not be always the most suitable way, but for `ServersSnapshot`, documenting it as _"A non-atomic snapshot ..."_ instead of as _"Provides access to a (non-atomic) snapshot ..."_ seems to be as OK as the above example. I can't on my own see why the proposed change makes the wording better. Could you please explain?